pkg/controller/volume/vacprotection/vac_protection_controller.go

```diff
@@ -0,0 +1,403 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vacprotection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	storagev1alpha1 "k8s.io/api/storage/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	storageinformers "k8s.io/client-go/informers/storage/v1alpha1"
+	clientset "k8s.io/client-go/kubernetes"
+	storagelisters "k8s.io/client-go/listers/storage/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller/volume/protectionutil"
+	"k8s.io/kubernetes/pkg/util/slice"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	vacNameKeyIndex = "volumeAttributesClassName"
+)
+
+// Controller is controller that adds and removes VACProtectionFinalizer
+// from VACs that are not used by any PV or PVC.
+type Controller struct {
+	client clientset.Interface
+
+	pvcSynced cache.InformerSynced
+	pvSynced  cache.InformerSynced
+	vacLister storagelisters.VolumeAttributesClassLister
+	vacSynced cache.InformerSynced
+
+	getPVsAssignedToVAC  func(vacName string) ([]*v1.PersistentVolume, error)
+	getPVCsAssignedToVAC func(vacName string) ([]*v1.PersistentVolumeClaim, error)
+
+	queue workqueue.RateLimitingInterface
+}
+
+// NewVACProtectionController returns a new *Controller.
+func NewVACProtectionController(logger klog.Logger,
+	client clientset.Interface,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	pvInformer coreinformers.PersistentVolumeInformer,
+	vacInformer storageinformers.VolumeAttributesClassInformer) (*Controller, error) {
+	c := &Controller{
+		client:    client,
+		pvcSynced: pvcInformer.Informer().HasSynced,
+		pvSynced:  pvInformer.Informer().HasSynced,
+		vacLister: vacInformer.Lister(),
+		vacSynced: vacInformer.Informer().HasSynced,
+		queue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "vacprotection"),
+	}
+
+	_, _ = vacInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.vacAddedUpdated(logger, obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			c.vacAddedUpdated(logger, new)
+		},
+	})
+
+	err := pvInformer.Informer().AddIndexers(cache.Indexers{vacNameKeyIndex: func(obj interface{}) ([]string, error) {
+		pv, ok := obj.(*v1.PersistentVolume)
+		if !ok {
+			return []string{}, nil
+		}
+		return getPVReferencedVACNames(pv), nil
+	}})
+	if err != nil {
+		return nil, fmt.Errorf("failed to add index to PV informer: %w", err)
+	}
+
+	pvIndexer := pvInformer.Informer().GetIndexer()
+	c.getPVsAssignedToVAC = func(vacName string) ([]*v1.PersistentVolume, error) {
```

Comment:
Nit: can this func be defined outside the NewVACProtectionController?

Reply:
If so, the controller has to add a field to hold the indexer.